# LAB | Feature Engineering

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In [1]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

**Check the shape of your data**

In [3]:
spaceship.shape

(8693, 14)

**Check for data types**

In [4]:
spaceship.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

**Check for missing values**

In [6]:
spaceship.isnull().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continouos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value. 

In [7]:
spaceship = spaceship.dropna()

In [8]:
spaceship.shape

(6606, 14)

- **Cabin** is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}

In [9]:
spaceship['Cabin'].value_counts()

Cabin
G/1476/S    7
E/13/S      7
C/137/S     7
G/734/S     7
B/11/S      7
           ..
E/233/S     1
E/209/P     1
G/548/S     1
D/108/P     1
B/153/P     1
Name: count, Length: 5305, dtype: int64

In [10]:
spaceship['Cabin'] = spaceship['Cabin'].str[0]
spaceship.head()

PassengerId HomePlanet CryoSleep Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False     B  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False     F  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False     A  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False     A  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False     F  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

- Drop PassengerId and Name

In [11]:
spaceship.drop(columns=['PassengerId', 'Name'], inplace=True)

In [12]:
spaceship.head()

HomePlanet CryoSleep Cabin  Destination   Age    VIP  RoomService  \
0     Europa     False     B  TRAPPIST-1e  39.0  False          0.0   
1      Earth     False     F  TRAPPIST-1e  24.0  False        109.0   
2     Europa     False     A  TRAPPIST-1e  58.0   True         43.0   
3     Europa     False     A  TRAPPIST-1e  33.0  False          0.0   
4      Earth     False     F  TRAPPIST-1e  16.0  False        303.0   

   FoodCourt  ShoppingMall     Spa  VRDeck  Transported  
0        0.0           0.0     0.0     0.0        False  
1        9.0          25.0   549.0    44.0         True  
2     3576.0           0.0  6715.0    49.0        False  
3     1283.0         371.0  3329.0   193.0        False  
4       70.0         151.0   565.0     2.0         True

- For non-numerical columns, do dummies.

In [18]:
spaceship = df = pd.get_dummies(spaceship, columns=['HomePlanet', 'Cabin', 'Destination'])

In [19]:
spaceship.head()

CryoSleep   Age    VIP  RoomService  FoodCourt  ShoppingMall     Spa  \
0     False  39.0  False          0.0        0.0           0.0     0.0   
1     False  24.0  False        109.0        9.0          25.0   549.0   
2     False  58.0   True         43.0     3576.0           0.0  6715.0   
3     False  33.0  False          0.0     1283.0         371.0  3329.0   
4     False  16.0  False        303.0       70.0         151.0   565.0   

   VRDeck  Transported  HomePlanet_Earth  ...  Cabin_B  Cabin_C  Cabin_D  \
0     0.0        False             False  ...     True    False    False   
1    44.0         True              True  ...    False    False    False   
2    49.0        False             False  ...    False    False    False   
3   193.0        False             False  ...    False    False    False   
4     2.0         True              True  ...    False    False    False   

   Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0    False    False    False    False                    False   
1    False     True    False    False                    False   
2    False    False    False    False                    False   
3    False    False    False    False                    False   
4    False     True    False    False                    False   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  
0                      False                     True  
1                      False                     True  
2                      False                     True  
3                      False                     True  
4                      False                     True  

[5 rows x 23 columns]

**Perform Train Test Split**

In [20]:
# Splitting the data into features (X) and target variable (y)
X = spaceship.drop(columns=['Transported'])
y = spaceship['Transported']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=22)

In [21]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (5284, 22)
X_test shape: (1322, 22)
y_train shape: (5284,)
y_test shape: (1322,)


**Model Selection**

In this exercise we will be using **KNN** as our predictive model.

In [22]:
from sklearn.neighbors import KNeighborsClassifier

In [23]:
knn = KNeighborsClassifier()

In [24]:
knn.fit(X_train, y_train)

KNeighborsClassifier()

- Evaluate your model's performance. Comment it

In [25]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

In [26]:
# Make predictions on the test set
y_pred = knn.predict(X_test)

# Compute accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Compute precision
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# Compute recall
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# Compute F1-score
f1 = f1_score(y_test, y_pred)
print("F1-score:", f1)

# Compute confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Accuracy: 0.7889561270801816
Precision: 0.8069908814589666
Recall: 0.7774524158125915
F1-score: 0.7919463087248322
Confusion Matrix:
[[512 127]
 [152 531]]


The K-Nearest Neighbors (KNN) classifier achieved an accuracy of about 78.90%, indicating that the model's predictions were correct for nearly 79% of the cases. The precision, measuring the accuracy of positive predictions, was approximately 80.70%, while recall, capturing the model's ability to find all positive instances, was around 77.75%. The F1-score, providing a balance between precision and recall, was approximately 79.19%.

In the confusion matrix, the model correctly classified 512 individuals as not being transported and 531 individuals as being transported. However, it misclassified 127 individuals as false negatives and 152 individuals as false positives. This suggests the model performs reasonably well but could be improved to reduce false predictions for better accuracy.

In [28]:
# Lets see if we can improve it

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

# Fit the scaler to the training data and transform the training data
X_train_scaled = scaler.fit_transform(X_train)

# Transform the test data using the scaler fitted on the training data
X_test_scaled = scaler.transform(X_test)

# Initialize the KNN classifier
knn2 = KNeighborsClassifier()

# Fit the KNN classifier to the scaled training data
knn2.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred = knn2.predict(X_test_scaled)

# Compute accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Compute precision
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# Compute recall
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# Compute F1-score
f1 = f1_score(y_test, y_pred)
print("F1-score:", f1)

# Compute confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Accuracy: 0.7760968229954615
Precision: 0.7927382753403933
Recall: 0.767203513909224
F1-score: 0.7797619047619048
Confusion Matrix:
[[502 137]
 [159 524]]


Scaling did not improve the model's performance. We would need to do additional analysis to improve the model even further.